Emby.Dlna/Ssdp/SsdpServer.cs

```diff
@@ -0,0 +1,624 @@
+        private readonly object _synchroniser;
+        private readonly ILogger _logger;
+        private readonly Hashtable _listeners;
+        private readonly Hashtable _senders;
+        private readonly INetworkManager _networkManager;
+        private readonly Dictionary<string, List<EventHandler<SsdpEventArgs>>> _events;
+        private readonly object _eventFireLock;
+        private Collection<IPObject> _interfaces;
+        private bool _running;
+        private bool _eventfire;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="SsdpServer"/> class.
+        /// </summary>
+        /// <param name="logger">The logger instance.<see cref="ILogger"/>.</param>
+        /// <param name="interfaces">Interfaces to use for the server.</param>
+        /// <param name="networkManager">The <see cref="INetworkManager"/> instance to use.</param>
+        private SsdpServer(ILogger logger, Collection<IPObject> interfaces, INetworkManager networkManager)
+        {
+            _logger = logger;
+            _eventFireLock = new object();
+            _synchroniser = new object();
+            _listeners = new Hashtable();
+            _senders = new Hashtable();
+            _events = new Dictionary<string, List<EventHandler<SsdpEventArgs>>>();
+            UdpSendCount = 2;
+            _interfaces = interfaces;
+            _networkManager = networkManager;
+        }
+
+        /// <summary>
+        /// Gets or sets the Host name to be used in SSDP packets.
+        /// </summary>
+        public static string HostName { get; set; } = $"OS/{Environment.OSVersion.VersionString} UPnP/1.0 RSSDP/1.0";
+
+        /// <summary>
+        /// Gets or sets a value indicating the number of times each udp packet should be sent.
+        /// </summary>
+        public int UdpSendCount { get; set; }
+
+        /// <summary>
+        /// Gets a value indicating whether detailed DNLA debug logging is active.
+        /// </summary>
+        public bool Tracing { get; internal set; }
+
+        /// <summary>
+        /// Gets a value indicating the tracing filter to be applied.
+        /// </summary>
+        public IPAddress? TracingFilter { get; internal set; }
+
+        /// <summary>
+        /// Gets or creates the singleton instance that is used by all objects.
+        /// </summary>
+        /// <param name="logger">Logger instance.</param>
+        /// <param name="interfaces">Interfaces to use for the server.</param>
+        /// <param name="networkManager">The <see cref="INetworkManager"/> instance to use.</param>
+        /// <returns>The SsdpServer singleton instance.</returns>
+        public static ISsdpServer GetOrCreateInstance(
+            ILogger logger,
+            Collection<IPObject> interfaces,
+            INetworkManager networkManager)
+        {
+            // As this class is used in multiple areas, we only want to create it once.
+            if (_instance == null)
```

Comment:
This can create multiple instances if called by multiple threads simultaneously. See https://csharpindepth.com/articles/singleton on how singletons can be implemented in a thread-safe manner. I would recommend the fourth version if the ssdp server isn't too expensive to create.

Reply:
That being said, is there any reason this can't be registered in DI as a singleton?

Reply:
Thanks.

Didn't add it to the DI as it allocates resources that may not need to be used. Will make it thread-safe instead.

Reply:
went with option 3 - as i don't want anything created before it actually needs to be.